### PR TITLE
RDF/DCAT support

### PR DESCRIPTION
**Pull request by noirbizarre**
This PR adds DCAT support:
- [x] Harvest RDF/DCAT endpoints
   - [x] RDFXML
   - [x] turtle
   - [x] JSON-LD
   - [x] n3
   - [x] nt
   - [x] guess license
   - [x] parse temporal coverage
   - [x] guess endpoint format from extension
   - [x] pagination support:
     - [x] hydra current (PartialCollection)
     - [x] hydra legacy (PaginatedCollection)
- [x] Expose RDF/DCAT endpoints
  - [x] JSON-LD context (`/context.jsonld`)
  - [x] catalog
    - [x] root content negociation (`/catalog`)
    - [x] RDFXML (`/catalog.xml`)
    - [x] turtle (`/catalog.ttl`)
    - [x] JSON-LD (`/catalog.json(ld)`)
    - [x] n3 (`/catalog.n3`)
    - [x] nt (`/catalog.nt`)
    - [x] pagination support
  - [x] dataset
    - [x] root content negociation (`/datasets/{id}/rdf`)
    - [x] html page content negociation (`{lang}/datasets/{id}/`)
    - [x] RDFXML (`/datasets/{id}/rdf.xml`)
    - [x] turtle (`/datasets/{id}/rdf.ttl`)
    - [x] JSON-LD (`/datasets/{id}/rdf.json(ld)`)
    - [x] n3 (`/datasets/{id}/rdf.n3`)
    - [x] nt (`/datasets/{id}/rdf.nt`)
    - [x] temporal coverage
  - [x] dataportal compliance (`/data.{json,xml,ttl}`)
- [x] links html pages to their rdf counterparts:
   - [x] home
   - [x] dataset
- [x] documenation
  - [x] harvesting documentation
  - [x] rdf documentation:
    - [x] endpoints
    - [x] fields mapping

These points can/should be done in other PRs:
- spatial coverage handling (parsing and exposition), idealy reusing geozones
- harvest content type negociation improvement (use headers...)
- allow search result to be available as rdf in addition to CSV (very easy)

### CODE REVIEW

**Comment by davidbgk** [comment]
Fixtures are huge 😱

File: udata/rdf.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    'application/ld+json': 'json-ld',
+    'application/json': 'json-ld',
+    'application/trig': 'trig',
+    # Available but no activated
+    # 'application/n-quads': 'nquads',
+    # 'text/xml': 'trix',
+}
+
+# Map formats to default used extensions
+RDF_EXTENSIONS = {
```
Comment (davidbgk) [inline udata/rdf.py]: Do we really need all these extensions? I would only activate jsonld for now. It impacts (at least) the number of links rendered in the template.

Comment (noirbizarre) [inline udata/rdf.py]: These are main RDF supported extensions, supporting them comes without extra programming cost and choice has been discussed and validated with @ColinMaudry.
Different format for different usage/tools.
We can't predict what usage or tools will be use to consume udata rdf (not so many tools support `JSON-LD`). In fact, I've been using different tools during that PR and not all of them support `json-ld`.
On the harvest part, we need to be able to harvest these formats: not all portals expose json-ld which is the younger format in the list (this is why this is the only one which requires an extra depdency to `rdflib`).
The DCAT spec is a RDF specification, not JSON-LD specific (`DCAT` vocabulary itself is not exposed as `JSON-LD` but only `xml` and `turtle`)

**Comment by noirbizarre** [comment]
I removed the fixtures that shouldn't be there.

**Comment by ColinMaudry** [comment]
Regarding content negotatiation, is it possible to negotiate an RDF format from `/datasets/{id}` ?

**Comment by noirbizarre** [comment]
Yes, still to be done (it's in the checklist).
The helpers are done, I just miss the view integration

**Comment by ColinMaudry** [comment]
Fixture = endpoint? I'm not sure of what you call a fixture.

**Comment by noirbizarre** [comment]
A fixture a a set of data required by the test(s).
In our case, fixtures are RDF files to simulate server response.

**Comment by noirbizarre** [comment]
Requires #967

**Comment by noirbizarre** [comment]
Requires #968

**Comment by noirbizarre** [comment]
Requires #969

File: docs/harvesting.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+                mime=faker.mime_type(category='text'),
+                format=faker.file_extension(category='text'),
+                filesize=faker.pyint()
+            ))
+
+        return dataset
+
+```
+
+You take a look at the [existing backends][backends-repository] to see exiting implementations.
```
Comment (vinyll) [inline docs/harvesting.md]: _You *may* take a look_

File: docs/harvesting.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
+# Harvesting
+
+Harvesting is the process of fetching of automatically remote metadata (ie. from other data portals or not)
+and store them into udata for being able to search them.
+
+## Vocabulary
+
+- **Backend**: designate a protocol implementation to harvest a remote endpoint.
```
Comment (noirbizarre) [inline docs/harvesting.md]: No, it doesn't mean the same thing ;)

File: docs/harvesting.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
+# Harvesting
+
+Harvesting is the process of fetching of automatically remote metadata (ie. from other data portals or not)
+and store them into udata for being able to search them.
```
Comment (davidbgk) [inline docs/harvesting.md]: `Harvesting is the process of automatically syncing remote metadata (i.e. from other data portals or not) and storing them into udata to be able to find remote data.`

Comment (noirbizarre) [inline docs/harvesting.md]: Wow, so many of typos in my sentence :scream: 
I'll fix that ASAP. But I think, we need to keep ``fetching` as `syncing` suggest it's bidirectionnal while harvesting is not.

File: docs/rdf.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    /dataset/{id}/rdf
+
+where `id` is dataset identifier on the udata instance.
+
+This URL performs content negociation and redirect to:
+
+    /dataset/{id}/rdf.{format}
+
+The dataset pages serve as identifier and perform content negociation too,
+so the following URL will all redirect to the same RDF endpoint:
```
Comment (davidbgk) [inline docs/rdf.md]: URLs

File: udata/core/dataset/rdf.py [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+
+def temporal_from_rdf(period_of_time):
+    '''Failsafe parsing of a temporal coverage'''
+    try:
+        if isinstance(period_of_time, Literal):
+            return temporal_from_literal(str(period_of_time))
+        elif isinstance(period_of_time, RdfResource):
+            return temporal_from_resource(period_of_time)
+    except:
+        # There is a lot of case where parsing could/should fail
```
Comment (davidbgk) [inline udata/core/dataset/rdf.py]: `are a lot of cases`

File: udata/core/dataset/views.py [hunk 7]
```diff
@@ -116,6 +136,33 @@ def resource_redirect(id):
...
+
+@blueprint.route('/<dataset:dataset>/rdf.<format>', localize=False)
+def rdf_format(dataset, format):
+    if not DatasetEditPermission(dataset).can():
+        if dataset.private:
+            abort(404)
+        elif dataset.deleted:
+            abort(410)
+
+    format = guess_format(format)
```
Comment (davidbgk) [inline udata/core/dataset/views.py]: These few lines are duplicated in `rdf_catalog_format`, mutualize?

File: udata/core/dataset/rdf.py [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+
+def temporal_from_rdf(period_of_time):
+    '''Failsafe parsing of a temporal coverage'''
+    try:
+        if isinstance(period_of_time, Literal):
+            return temporal_from_literal(str(period_of_time))
+        elif isinstance(period_of_time, RdfResource):
+            return temporal_from_resource(period_of_time)
+    except:
+        # There are a lot of cases where parsing could/should fail
```
Comment (davidbgk) [inline udata/core/dataset/rdf.py]: Is that worth logging it for future improvements?

File: udata/rdf.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+class UDataContext(Context):
+    '''
+    An hackish way to serialize context as a root relative URL.
+
+    Exploit this issue https://github.com/RDFLib/rdflib-jsonld/issues/37
+    and the fact that this method is used to render the context in the
+    resulting JSON-LD.
+
+    See: https://github.com/RDFLib/rdflib-jsonld/blob/master/rdflib_jsonld/serializer.py#L101-L103
+    '''  # noqa
```
Comment (davidbgk) [inline udata/rdf.py]: If you put only a URL on one line, the linter shouldn't complain even without a # noqa comment. At least mine 😉

Comment (noirbizarre) [inline udata/rdf.py]: Mine is crying without :cry:

**Pull request merged**